### PR TITLE
Annotation model

### LINKED ISSUE #138: My Annotations
Create both my and their annotations pages (`/my_account/audio_events`, `/user_accounts/:id/audio_events`)

### LINKED ISSUE #563: Create toRelative pipe
Standardize how `Durations` and `DateTime`s are converted to relative time by creating a toRelative pipe which handles both. This should allow the following operations to occur in the angular html code:

```typescript
const duration = new Duration();
const date = new DateTime.now();
```
```html
<p>{{ duration | toRelative:'short' }}</p>
<p>{{ date | toRelative:'long' }}</p>
```

### PR DESCRIPTION
**Pull request by Allcharles**
# Implemented Annotation Pages

Created the my annotations and their annotations pages

## Changes

- Implemented My Annotation Page
- Implemented Their Annotation Page
- Modified how the Association Decorators function slightly to reduce the number of excess queries, and improve type checking
- Created pipe for displaying loading animation
- Created pipe for testing if a model is resolved
- Created pipe for converting `Duration`s and `DateTime`s to strings

## Problems

- [x] Need to create dummy links for Play and Visualize buttons
- [x] Need to update unit tests
- [x] Modifications to UnresolvedModel.many should be reverted. Checking for a list of UnresolvedModels can be done using a comparison of pointers

## Issues

Fixes: #138
Fixes: #563 

### Page Loading

temp

### My Annotations Page

image

### Their Annotations Page

image

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing

### CODE REVIEW

File: src/app/helpers/tableTemplate/pagedTableTemplate.ts [hunk 1]
```diff
@@ -152,7 +154,7 @@ export abstract class PagedTableTemplate<T, M extends AbstractModel>
       );
   }
 
-  protected apiAction(filters: Filters<M>, args: AbstractModel[]) {
+  protected apiAction(filters: Filters<M>, args?: AbstractModel[]) {
```
Comment (Allcharles) [inline src/app/helpers/tableTemplate/pagedTableTemplate.ts]: I suspect this should instead be `protected apiAction(filters: Filters<M>, args: AbstractModel[] = []) {`

File: src/app/components/profile/pages/annotations/annotations.component.html [hunk 2]
```diff
@@ -21,11 +22,35 @@ <h1>
   >
     <ngx-datatable-column name="Site" [sortable]="false">
       <ng-template let-value="value" ngx-datatable-cell-template>
-        <a [routerLink]="[value.viewUrl]">{{ value.name }}</a>
+        {{ value.audioRecording?.site?.name || "(loading)" }}
```
Comment (atruskie) [inline src/app/components/profile/pages/annotations/annotations.component.html]: make a loading standard pipe?

File: src/app/components/profile/pages/annotations/annotations.component.html [hunk 3]
```diff
@@ -21,11 +22,35 @@ <h1>
...
+      <ng-template let-value="value" ngx-datatable-cell-template>
+        {{ value }}
       </ng-template>
     </ngx-datatable-column>
-    <ngx-datatable-column name="Updated"> </ngx-datatable-column>
     <ngx-datatable-column name="Tags" [sortable]="false">
+      <ng-template let-value="value" ngx-datatable-cell-template>
+        <ng-container *ngFor="let tag of value.tags">
+          <span class="badge badge-primary mr-1">
+            {{ tag.text || "(loading)" }}
```
Comment (atruskie) [inline src/app/components/profile/pages/annotations/annotations.component.html]: pipe?

File: src/app/helpers/tableTemplate/pagedTableTemplate.ts (hunk 1 above)
Comment (atruskie) [reply src/app/helpers/tableTemplate/pagedTableTemplate.ts]: I don't think it matters. According to the spec it should resolve the same when its spread.

File: src/app/models/AbstractModel.ts [hunk 4]
```diff
@@ -101,7 +101,7 @@ export abstract class AbstractModel {
 
 export class UnresolvedModel extends AbstractModel {
   private static readonly model = Object.freeze(new UnresolvedModel());
-  private static readonly models = [];
+  private static readonly models = [UnresolvedModel.model];
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]: should freexe this array to so it can't be mutated

File: src/app/models/AssociationDecorators.ts [hunk 5]
```diff
@@ -1,205 +1,224 @@
...
- * Associate models with list of IDs
- * @param serviceToken Injection token for API Service
- * @param modelIdentifier Property to read IDs from
- * @param modelPrimaryKey Keys to match additional ids against
+ * Abstract model parameter decorator which automates the process
+ * of retrieving models linked by a group of ids in the parent model.
+ *
+ * @param serviceToken Injection token for API service used to retrieve the child models
+ * @param identifierKeys Parent model key used to retrieve the list of ids for the child models
+ * @param childIdentifier Primary key of the child model
```
Comment (atruskie) [inline src/app/models/AssociationDecorators.ts]:
```suggestion
 * @param childIdentifier field used to filter child models by, typically ìd`
```

File: src/app/models/AudioEvent.ts [hunk 6]
```diff
@@ -50,14 +65,23 @@ export class AudioEvent extends AbstractModel implements IAudioEvent {
...
+  @HasOne<AudioEvent, AudioRecording>(AUDIO_RECORDING, "audioRecordingId")
   public audioRecording?: AudioRecording;
+  @HasMany<AudioEvent, Tag>(TAG, "tagIds")
+  public tags?: Tag[];
 
   constructor(audioEvent: IAudioEvent, injector?: Injector) {
     super(audioEvent, injector);
+    this.taggings = (audioEvent.taggings as ITagging[]).map(
+      (tagging) => new Tagging(tagging, injector)
+    );
```
Comment (atruskie) [inline src/app/models/AudioEvent.ts]: As an FYI this pattern is very rare. Our API advice acturally states whole nested objects should never be serialised

**Review by atruskie — APPROVED** [review]

File: src/app/models/AssociationDecorators.ts [hunk 7]
```diff
@@ -45,7 +45,7 @@ export function Deleter<Parent extends AbstractModel & { deleterId?: Id }>() {
  *
  * @param serviceToken Injection token for API service used to retrieve the child models
  * @param identifierKeys Parent model key used to retrieve the list of ids for the child models
- * @param childIdentifier Primary key of the child model
+ * @param childIdentifier field used to filter child models by, typically ìd`
```
Comment (atruskie) [inline src/app/models/AssociationDecorators.ts:48]: Looks like by suggestion was bad: I has the international keyboard on!
```suggestion
 * @param childIdentifier field used to filter child models by, typically `id`
```

File: src/app/components/profile/pages/annotations/annotations.component.html [hunk 8]
```diff
@@ -22,7 +22,9 @@ <h1>
   >
     <ngx-datatable-column name="Site" [sortable]="false">
       <ng-template let-value="value" ngx-datatable-cell-template>
-        {{ value.audioRecording?.site?.name || "(loading)" }}
+        <span
+          [innerHTML]="value.audioRecording?.site?.name | spinner: 'sm'"
```
Comment (atruskie) [inline src/app/components/profile/pages/annotations/annotations.component.html]: when you get to the point of modifying html, it is probably time for a component?

**Review by atruskie — APPROVED** [review]
I think maybe the spinner pipe is best collapsed into the spinner component we have.

Other than that, great!

**Pull request merged**